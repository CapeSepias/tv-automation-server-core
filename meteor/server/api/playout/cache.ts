import { ReadonlyDeep } from 'type-fest'
import _ from 'underscore'
import { PartInstance, DBPartInstance, PartInstances } from '../../../lib/collections/PartInstances'
import { Part, DBPart, Parts } from '../../../lib/collections/Parts'
import { PeripheralDevice, PeripheralDevices } from '../../../lib/collections/PeripheralDevices'
import { PieceInstance, PieceInstances } from '../../../lib/collections/PieceInstances'
import {
	RundownPlaylist,
	DBRundownPlaylist,
	RundownPlaylistId,
	RundownPlaylists,
} from '../../../lib/collections/RundownPlaylists'
import { Rundown, DBRundown, Rundowns } from '../../../lib/collections/Rundowns'
import { Segment, DBSegment, Segments } from '../../../lib/collections/Segments'
import { Studio, StudioId, Studios } from '../../../lib/collections/Studios'
import { Timeline, TimelineComplete } from '../../../lib/collections/Timeline'
import { waitForPromise } from '../../../lib/lib'
import { ActivationCache, getActivationCache } from '../../cache/ActivationCache'
import { DbCacheReadCollection, DbCacheWriteCollection } from '../../cache/CacheCollection'
import { DbCacheReadObject, DbCacheWriteObject } from '../../cache/CacheObject'
import { CacheBase } from '../../cache/DatabaseCaches'
import { profiler } from '../profiler'
import { removeRundownPlaylistFromDb } from '../rundownPlaylist'
import { CacheForStudioBase } from '../studio/cache'

export abstract class CacheForPlayoutPreInit extends CacheBase {
	public readonly isPlayout = true

	public readonly activationCache: ActivationCache

	public readonly Studio: DbCacheReadObject<Studio, Studio>
	public readonly PeripheralDevices: DbCacheReadCollection<PeripheralDevice, PeripheralDevice>

	public readonly Playlist: DbCacheWriteObject<RundownPlaylist, DBRundownPlaylist>
	public readonly Rundowns: DbCacheWriteCollection<Rundown, DBRundown> // TODO DbCacheReadCollection??

	protected constructor(studioId: StudioId, playlistId: RundownPlaylistId) {
		super()

		this.activationCache = getActivationCache(studioId, playlistId)

		this.Studio = new DbCacheReadObject(Studios, false)
		this.PeripheralDevices = new DbCacheReadCollection(PeripheralDevices)

		this.Playlist = new DbCacheWriteObject(RundownPlaylists, false)
		this.Rundowns = new DbCacheWriteCollection(Rundowns)
	}

	protected async preInit(tmpPlaylist: ReadonlyDeep<RundownPlaylist>) {
		await Promise.all([
			this.Playlist._initialize(tmpPlaylist._id),
			this.Rundowns.prepareInit({ playlistId: tmpPlaylist._id }, true),
		])

		const rundowns = this.Rundowns.findFetch()
		await this.activationCache.initialize(tmpPlaylist, rundowns)

		this.Studio._fromDoc(this.activationCache.getStudio())
		await this.PeripheralDevices.prepareInit(
			() => waitForPromise(this.activationCache.getPeripheralDevices()),
			true
		)
	}
}

export class CacheForPlayout extends CacheForPlayoutPreInit implements CacheForStudioBase {
	private toBeRemoved: boolean = false

	public readonly Timeline: DbCacheWriteCollection<TimelineComplete, TimelineComplete>

	public readonly Segments: DbCacheReadCollection<Segment, DBSegment>
	public readonly Parts: DbCacheWriteCollection<Part, DBPart> // TODO DbCacheReadCollection
	public readonly PartInstances: DbCacheWriteCollection<PartInstance, DBPartInstance>
	public readonly PieceInstances: DbCacheWriteCollection<PieceInstance, PieceInstance>

	private constructor(studioId: StudioId, playlistId: RundownPlaylistId) {
		super(studioId, playlistId)

		this.Timeline = new DbCacheWriteCollection<TimelineComplete, TimelineComplete>(Timeline)

		this.Segments = new DbCacheReadCollection<Segment, DBSegment>(Segments)
		this.Parts = new DbCacheWriteCollection<Part, DBPart>(Parts)

		this.PartInstances = new DbCacheWriteCollection<PartInstance, DBPartInstance>(PartInstances)
		this.PieceInstances = new DbCacheWriteCollection<PieceInstance, PieceInstance>(PieceInstances)
	}

	static async create(tmpPlaylist: ReadonlyDeep<RundownPlaylist>): Promise<CacheForPlayout> {
		const res = new CacheForPlayout(tmpPlaylist.studioId, tmpPlaylist._id)

		await res.preInit(tmpPlaylist)

		return res
	}

	async initContent(): Promise<void> {
		const playlist = this.Playlist.doc

		const ps: Promise<any>[] = []

		const rundownIds = this.Rundowns.findFetch().map((r) => r._id)

		const selectedPartInstanceIds = _.compact([
			playlist.currentPartInstanceId,
			playlist.nextPartInstanceId,
			playlist.previousPartInstanceId,
		])

		ps.push(this.Segments.prepareInit({ rundownId: { $in: rundownIds } }, true)) // TODO - omit if we cant or are unlikely to change the current part
		ps.push(this.Parts.prepareInit({ rundownId: { $in: rundownIds } }, true)) // TODO - omit if we cant or are unlikely to change the current part

		ps.push(
			this.PartInstances.prepareInit(
				{
					playlistActivationId: playlist.activationId,
					rundownId: { $in: rundownIds },
					reset: { $ne: true },
				},
				true
			)
		)

		ps.push(
			this.PieceInstances.prepareInit(
				{
					playlistActivationId: playlist.activationId,
					rundownId: { $in: rundownIds },
					partInstanceId: { $in: selectedPartInstanceIds },
					reset: { $ne: true },
				},
				true
			)
		)

		await Promise.all(ps)

		// This will be needed later, but we will do some other processing first
		// TODO-CACHE what happens if this errors? where should that go?
		Promise.all([this.Timeline.prepareInit({ studioId: playlist.studioId }, true)])
	}

	removePlaylist() {
		this.toBeRemoved = true
	}

	discardChanges() {
		this.toBeRemoved = false
		super.discardChanges()
	}

	async saveAllToDatabase() {
		if (this.toBeRemoved) {
			const span = profiler.startSpan('CacheForPlayout.saveAllToDatabase')
			this._abortActiveTimeout()

			// TODO - run any of the defers?

			waitForPromise(removeRundownPlaylistFromDb(this.Playlist.doc._id))

			span?.end()
		} else {
			return super.saveAllToDatabase()
		}
	}
}
