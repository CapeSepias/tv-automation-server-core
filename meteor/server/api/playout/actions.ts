import { Meteor } from 'meteor/meteor'
import * as _ from 'underscore'
import { logger } from '../../logging'
import { Rundown, Rundowns, RundownHoldState } from '../../../lib/collections/Rundowns'
import { PeripheralDevices, PeripheralDevice } from '../../../lib/collections/PeripheralDevices'
import { PeripheralDeviceAPI } from '../../../lib/api/peripheralDevice'
import { getCurrentTime, getRandomId, waitForPromise } from '../../../lib/lib'
import { loadShowStyleBlueprint } from '../blueprints/cache'
import { RundownContext } from '../blueprints/context'
import {
	setNextPart,
	onPartHasStoppedPlaying,
	selectNextPart,
	getSelectedPartInstancesFromCache,
	getAllOrderedPartsFromPlayoutCache,
} from './lib'
import { updateTimeline } from './timeline'
import { IngestActions } from '../ingest/actions'
import { getActiveRundownPlaylistsInStudioFromDb } from './studio'
import { CacheForPlayout, CacheForStudio } from '../../cache/DatabaseCaches'
import { profiler } from '../profiler'

export function activateRundownPlaylist(cache: CacheForPlayout, rehearsal: boolean): void {
	{
		const playlist = cache.Playlist.doc
		logger.info('Activating rundown ' + playlist._id + (rehearsal ? ' (Rehearsal)' : ''))

		rehearsal = !!rehearsal
		// if (rundown.active && !rundown.rehearsal) throw new Meteor.Error(403, `Rundown "${rundown._id}" is active and not in rehersal, cannot reactivate!`)

		const studio = cache.Studio.doc

		const anyOtherActiveRundowns = getActiveRundownPlaylistsInStudioFromDb(studio._id, playlist._id)

		if (anyOtherActiveRundowns.length) {
			// logger.warn('Only one rundown can be active at the same time. Active rundowns: ' + _.map(anyOtherActiveRundowns, rundown => rundown._id))
			throw new Meteor.Error(
				409,
				'Only one rundown can be active at the same time. Active rundown playlists: ' +
					_.map(anyOtherActiveRundowns, (playlist) => playlist._id),
				JSON.stringify(_.map(anyOtherActiveRundowns, (playlist) => playlist._id))
			)
		}

		cache.Playlist.update({
			$set: {
				active: true,
				rehearsal: rehearsal,
				activeInstanceId: getRandomId(),
			},
		})
	}

	const activePlaylist = cache.Playlist.doc

	// Re-Initialize the ActivationCache now when the rundownPlaylist is active
	const rundownsInPlaylist = cache.Rundowns.findFetch()
	waitForPromise(cache.activationCache.initialize(activePlaylist, rundownsInPlaylist))

	if (!activePlaylist.nextPartInstanceId) {
		const firstPart = selectNextPart(activePlaylist, null, getAllOrderedPartsFromPlayoutCache(cache))
		setNextPart(cache, firstPart ? firstPart.part : null)
	} else {
		const nextPartInstance = cache.PartInstances.findOne(activePlaylist.nextPartInstanceId)
		if (!nextPartInstance)
			throw new Meteor.Error(404, `Could not find nextPartInstance "${activePlaylist.nextPartInstanceId}"`)
		const rundown = cache.Rundowns.findOne(nextPartInstance.rundownId)
		if (!rundown) throw new Meteor.Error(404, `Could not find rundown "${nextPartInstance.rundownId}"`)

		cache.defer(() => {
			if (!rundown) return // if the proper rundown hasn't been found, there's little point doing anything else
			const showStyleCompound = waitForPromise(cache.activationCache.getShowStyleCompound(rundown))
			const { blueprint } = loadShowStyleBlueprint(showStyleCompound)

			const context = new RundownContext(cache.Studio.doc, rundown, showStyleCompound, undefined)
			context.wipeCache()

			if (blueprint.onRundownActivate) {
				Promise.resolve(blueprint.onRundownActivate(context)).catch(logger.error)
			}
		})
	}

	updateTimeline(cache)
}
export function deactivateRundownPlaylist(cache: CacheForPlayout): void {
	const rundown = deactivateRundownPlaylistInner(cache)

	updateTimeline(cache)

	cache.defer(() => {
		if (rundown) {
			const showStyleBase = waitForPromise(cache.activationCache.getShowStyleCompound(rundown))
			const { blueprint } = loadShowStyleBlueprint(showStyleBase)

			if (blueprint.onRundownDeActivate) {
				const context = new RundownContext(cache.Studio.doc, rundown, showStyleBase, undefined)
				Promise.resolve(blueprint.onRundownDeActivate(context)).catch(logger.error)
			}
		}
	})
}
export function deactivateRundownPlaylistInner(cache: CacheForPlayout): Rundown | undefined {
	const span = profiler.startSpan('deactivateRundownPlaylistInner')
	logger.info(`Deactivating rundown playlist "${cache.Playlist.doc._id}"`)

	const { currentPartInstance, nextPartInstance } = getSelectedPartInstancesFromCache(cache)

	let rundown: Rundown | undefined
	if (currentPartInstance) {
		rundown = cache.Rundowns.findOne(currentPartInstance.rundownId)
		// defer so that an error won't prevent deactivate
		cache.deferAfterSave(() => {
			// This is low-prio, deferring
			Meteor.setTimeout(() => {
				if (rundown) {
					IngestActions.notifyCurrentPlayingPart(rundown, null)
				} else {
					logger.error(
						`Could not find owner Rundown "${currentPartInstance.rundownId}" of PartInstance "${currentPartInstance._id}"`
					)
				}
			}, LOW_PRIO_DEFER_TIME)
		})
	} else {
		if (nextPartInstance) {
			rundown = cache.Rundowns.findOne(nextPartInstance.rundownId)
		}
	}

	if (currentPartInstance) onPartHasStoppedPlaying(cache, currentPartInstance, getCurrentTime())

	cache.Playlist.update({
		$set: {
			active: false,
			previousPartInstanceId: null,
			currentPartInstanceId: null,
			holdState: RundownHoldState.NONE,
		},
	})
	setNextPart(cache, null)

	if (currentPartInstance) {
		cache.PartInstances.update(currentPartInstance._id, {
			$set: {
				'timings.takeOut': getCurrentTime(),
			},
		})
	}
	if (span) span.end()
	return rundown
}
/**
 * Prepares studio before a broadcast is about to start
 * @param studio
 * @param okToDestoryStuff true if we're not ON AIR, things might flicker on the output
 */
export function prepareStudioForBroadcast(cache: CacheForPlayout, okToDestoryStuff: boolean): void {
	const rundownPlaylistToBeActivated = cache.Playlist.doc
	if (!rundownPlaylistToBeActivated.studioId)
		throw new Meteor.Error(500, `Playlist "${rundownPlaylistToBeActivated._id}" has no studioId!`)

	logger.info('prepareStudioForBroadcast ' + rundownPlaylistToBeActivated.studioId)

	const playoutDevices: Array<Pick<PeripheralDevice, '_id'>> = PeripheralDevices.find(
		{
			studioId: rundownPlaylistToBeActivated.studioId,
			type: PeripheralDeviceAPI.DeviceType.PLAYOUT,
		},
		{
			fields: { _id: 1 },
		}
	).fetch()

	_.each(playoutDevices, (device: PeripheralDevice) => {
		PeripheralDeviceAPI.executeFunction(
			device._id,
			(err) => {
				if (err) {
					logger.error(err)
				} else {
					logger.info('devicesMakeReady OK')
				}
			},
			'devicesMakeReady',
			okToDestoryStuff,
			rundownPlaylistToBeActivated._id
		)
	})
}
/**
 * Makes a studio "stand down" after a broadcast
 * @param studio
 * @param okToDestoryStuff true if we're not ON AIR, things might flicker on the output
 */
export function standDownStudio(cache: CacheForPlayout | CacheForStudio, okToDestoryStuff: boolean): void {
	const studio = cache.Studio.doc
	logger.info('standDownStudio ' + studio._id)

	const playoutDevices: Array<Pick<PeripheralDevice, '_id'>> = PeripheralDevices.find(
		{
			studioId: studio._id,
			type: PeripheralDeviceAPI.DeviceType.PLAYOUT,
		},
		{
			fields: { _id: 1 },
		}
	).fetch()

	_.each(playoutDevices, (device: PeripheralDevice) => {
		PeripheralDeviceAPI.executeFunction(
			device._id,
			(err) => {
				if (err) {
					logger.error(err)
				} else {
					logger.info('devicesStandDown OK')
				}
			},
			'devicesStandDown',
			okToDestoryStuff
		)
	})
}
