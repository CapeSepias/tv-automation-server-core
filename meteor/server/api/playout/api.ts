import { Methods, setMeteorMethods } from '../../methods'
import { PlayoutAPI } from '../../../lib/api/playout'
import { ServerPlayoutAPI } from './playout'
import { getCurrentTime } from '../../../lib/lib'
import { logger } from '../../logging'
import { RundownPlaylistId } from '../../../lib/collections/RundownPlaylists'
import { PartInstanceId } from '../../../lib/collections/PartInstances'
import { PartId } from '../../../lib/collections/Parts'
import { PieceId } from '../../../lib/collections/Pieces'
import { StudioId } from '../../../lib/collections/Studios'
import { PieceInstanceId } from '../../../lib/collections/PieceInstances'

let methods: Methods = {}
methods[PlayoutAPI.methods.rundownPrepareForBroadcast] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.prepareRundownPlaylistForBroadcast(playlistId)
}
methods[PlayoutAPI.methods.rundownResetRundown] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.resetRundownPlaylist(playlistId)
}
methods[PlayoutAPI.methods.rundownResetAndActivate] = (playlistId: RundownPlaylistId, rehearsal?: boolean) => {
	return ServerPlayoutAPI.resetAndActivateRundownPlaylist(playlistId, rehearsal)
}
methods[PlayoutAPI.methods.rundownActivate] = (playlistId: RundownPlaylistId, rehearsal: boolean) => {
	return ServerPlayoutAPI.activateRundownPlaylist(playlistId, rehearsal)
}
methods[PlayoutAPI.methods.rundownDeactivate] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.deactivateRundownPlaylist(playlistId)
}
methods[PlayoutAPI.methods.reloadRundownPlaylistData] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.reloadRundownPlaylistData(playlistId)
}
methods[PlayoutAPI.methods.pieceTakeNow] = (playlistId: RundownPlaylistId, partInstanceId: PartInstanceId, pieceInstanceIdOrPieceIdToCopy: PieceInstanceId | PieceId) => {
	return ServerPlayoutAPI.pieceTakeNow(playlistId, partInstanceId, pieceInstanceIdOrPieceIdToCopy)
}
methods[PlayoutAPI.methods.rundownTake] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.takeNextPart(playlistId)
}
methods[PlayoutAPI.methods.rundownTogglePartArgument] = (playlistId: RundownPlaylistId, partInstanceId: PartInstanceId, property: string, value: string) => {
	return ServerPlayoutAPI.rundownTogglePartArgument(playlistId, partInstanceId, property, value)
}
methods[PlayoutAPI.methods.rundownSetNext] = (playlistId: RundownPlaylistId, partId: PartId, timeOffset?: number | undefined) => {
	return ServerPlayoutAPI.setNextPart(playlistId, partId, true, timeOffset)
}
methods[PlayoutAPI.methods.rundownMoveNext] = (playlistId: RundownPlaylistId, horisontalDelta: number, verticalDelta: number) => {
	return ServerPlayoutAPI.moveNextPart(playlistId, horisontalDelta, verticalDelta, true)
}
methods[PlayoutAPI.methods.rundownActivateHold] = (playlistId: RundownPlaylistId) => {
	return ServerPlayoutAPI.activateHold(playlistId)
}
methods[PlayoutAPI.methods.rundownDisableNextPiece] = (rundownPlaylistId: RundownPlaylistId, undo?: boolean) => {
	return ServerPlayoutAPI.disableNextPiece(rundownPlaylistId, undo)
}
// methods[PlayoutAPI.methods.partPlaybackStartedCallback] = (rundownId: RundownId, partId: PartId, startedPlayback: number) => {
// 	return ServerPlayoutAPI.onPartPlaybackStarted(rundownId, partId, startedPlayback)
// }
// methods[PlayoutAPI.methods.piecePlaybackStartedCallback] = (rundownId: RundownId, pieceId: PieceId, startedPlayback: number) => {
// 	return ServerPlayoutAPI.onPiecePlaybackStarted(rundownId, pieceId, startedPlayback)
// }
methods[PlayoutAPI.methods.segmentAdLibPieceStart] = (rundownPlaylistId: RundownPlaylistId, partInstanceId: PartInstanceId, pieceId: PieceId, queue: boolean) => {
	return ServerPlayoutAPI.segmentAdLibPieceStart(rundownPlaylistId, partInstanceId, pieceId, queue)
}
methods[PlayoutAPI.methods.rundownBaselineAdLibPieceStart] = (rundownPlaylistId: RundownPlaylistId, partInstanceId: PartInstanceId, pieceId: PieceId, queue: boolean) => {
	return ServerPlayoutAPI.rundownBaselineAdLibPieceStart(rundownPlaylistId, partInstanceId, pieceId, queue)
}
methods[PlayoutAPI.methods.sourceLayerOnPartStop] = (rundownPlaylistId: RundownPlaylistId, partInstanceId: PartInstanceId, sourceLayerId: string) => {
	return ServerPlayoutAPI.sourceLayerOnPartStop(rundownPlaylistId, partInstanceId, sourceLayerId)
}
// methods[PlayoutAPI.methods.timelineTriggerTimeUpdateCallback] = (timelineObjId: string, time: number) => {
// 	return ServerPlayoutAPI.timelineTriggerTimeUpdateCallback(timelineObjId, time)
// }
methods[PlayoutAPI.methods.sourceLayerStickyPieceStart] = (playlistId: RundownPlaylistId, sourceLayerId: string) => {
	return ServerPlayoutAPI.sourceLayerStickyPieceStart(playlistId, sourceLayerId)
}
methods[PlayoutAPI.methods.updateStudioBaseline] = (studioId: StudioId) => {
	return ServerPlayoutAPI.updateStudioBaseline(studioId)
}
methods[PlayoutAPI.methods.shouldUpdateStudioBaseline] = (studioId: StudioId) => {
	return ServerPlayoutAPI.shouldUpdateStudioBaseline(studioId)
}
// Apply methods:
setMeteorMethods(methods)

// Temporary methods
setMeteorMethods({
	'debug__printTime': () => {
		let now = getCurrentTime()
		logger.debug(new Date(now))
		return now
	},
})
