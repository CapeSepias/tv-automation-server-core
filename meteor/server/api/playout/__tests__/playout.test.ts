import { Meteor } from 'meteor/meteor'
import '../../../../__mocks__/_extendJest'
import { testInFiber } from '../../../../__mocks__/helpers/jest'
import { fixSnapshot } from '../../../../__mocks__/helpers/snapshot'
import { mockupCollection } from '../../../../__mocks__/helpers/lib'
import { setupDefaultStudioEnvironment, DefaultEnvironment, setupDefaultRundown, setupMockPeripheralDevice } from '../../../../__mocks__/helpers/database'
import { Rundowns, Rundown } from '../../../../lib/collections/Rundowns'
import '../api'
import { Timeline as OrgTimeline } from '../../../../lib/collections/Timeline'
import { ServerPlayoutAPI } from '../playout'
import { deactivate } from '../../userActions'
import { PeripheralDevice } from '../../../../lib/collections/PeripheralDevices'
import { PeripheralDeviceCommands } from '../../../../lib/collections/PeripheralDeviceCommands'
import { PeripheralDeviceAPI } from '../../../../lib/api/peripheralDevice'
import { Pieces } from '../../../../lib/collections/Pieces'
import { AdLibPieces } from '../../../../lib/collections/AdLibPieces'

const Timeline = mockupCollection(OrgTimeline)

describe('Playout API', () => {
	let env: DefaultEnvironment
	let playoutDevice: PeripheralDevice

	function getPeripheralDeviceCommands (playoutDevice: PeripheralDevice) {
		return PeripheralDeviceCommands.find({ deviceId: playoutDevice._id }, { sort: { time: 1 } }).fetch()
	}
	function clearPeripheralDeviceCommands (playoutDevice: PeripheralDevice) {
		return PeripheralDeviceCommands.remove({ deviceId: playoutDevice._id })
	}
	function getAllRundownData (rundown: Rundown) {
		return {
			parts: rundown.getParts(),
			segments: rundown.getSegments(),
			rundown: Rundowns.findOne(rundown._id) as Rundown,
			pieces: Pieces.find({ rundown: rundown._id }, { sort: { _id: 1 } }).fetch(),
			adLibPieces: AdLibPieces.find({ rundown: rundown._id }, { sort: { _id: 1 } }).fetch()
		}
	}
	beforeEach(() => {
		env = setupDefaultStudioEnvironment()
		playoutDevice = setupMockPeripheralDevice(
			PeripheralDeviceAPI.DeviceCategory.PLAYOUT,
			PeripheralDeviceAPI.DeviceType.PLAYOUT,
			PeripheralDeviceAPI.SUBTYPE_PROCESS,
			env.studio
		)
		// @ts-ignore
		Timeline.insert.mockClear()
		// @ts-ignore
		Timeline.upsert.mockClear()
		// @ts-ignore
		Timeline.update.mockClear()
	})
	testInFiber('Basic rundown control', () => {
		const rundownId0 = setupDefaultRundown(env)
		expect(rundownId0).toBeTruthy()

		const getRundown0 = () => {
			return Rundowns.findOne(rundownId0) as Rundown
		}
		const parts = getRundown0().getParts()

		expect(getRundown0()).toMatchObject({
			active: false,
			rehearsal: false
		})

		expect(Timeline.insert).not.toHaveBeenCalled()
		expect(Timeline.upsert).not.toHaveBeenCalled()
		expect(Timeline.update).not.toHaveBeenCalled()

		ServerPlayoutAPI.resetRundown(rundownId0)
		const orgRundownData = getAllRundownData(getRundown0())

		// Prepare and activate in rehersal:
		ServerPlayoutAPI.activateRundown(rundownId0, false)
		expect(getRundown0()).toMatchObject({
			active: true,
			rehearsal: false,
			currentPartId: null,
			nextPartId: parts[0]._id,
		})

		expect(Timeline.insert).toHaveBeenCalled()
		expect(Timeline.upsert).toHaveBeenCalled()
		expect(Timeline.update).toHaveBeenCalled()
		Timeline.mockClear()

		// Take the first Part:
		ServerPlayoutAPI.takeNextPart(rundownId0)
		expect(getRundown0()).toMatchObject({
			currentPartId: parts[0]._id,
			nextPartId: parts[1]._id,
		})

		expect(Timeline.insert).toHaveBeenCalled()
		expect(Timeline.upsert).toHaveBeenCalled()
		expect(Timeline.update).toHaveBeenCalled()
		Timeline.mockClear()

		expect(fixSnapshot(Timeline.find().fetch())).toMatchSnapshot()
		expect(fixSnapshot(getRundown0())).toMatchSnapshot()


		// Deactivate rundown:
		ServerPlayoutAPI.deactivateRundown(rundownId0)
		expect(getRundown0()).toMatchObject({
			active: false,
			currentPartId: null,
			nextPartId: null
		})

		expect(fixSnapshot(Timeline.find().fetch())).toMatchSnapshot()
		expect(fixSnapshot(getRundown0())).toMatchSnapshot()

		expect(Timeline.insert).toHaveBeenCalled()
		expect(Timeline.upsert).toHaveBeenCalled()
		expect(Timeline.update).toHaveBeenCalled()

		// lastly: reset rundown
		ServerPlayoutAPI.resetRundown(rundownId0)

		// Verify that the data is back to as it was before any of the operations:
		const rundownData = getAllRundownData(getRundown0())
		expect(rundownData).toEqual(orgRundownData)

	})
	testInFiber('prepareRundownForBroadcast', () => {
		const rundownId0 = setupDefaultRundown(env)
		expect(rundownId0).toBeTruthy()

		const getRundown0 = () => {
			return Rundowns.findOne(rundownId0) as Rundown
		}

		expect(getRundown0()).toMatchObject({
			active: false,
			rehearsal: false
		})

		expect(getPeripheralDeviceCommands(playoutDevice)).toHaveLength(0)

		// Prepare and activate in rehersal:
		ServerPlayoutAPI.prepareRundownForBroadcast(rundownId0)

		expect(getRundown0()).toMatchObject({
			active: true,
			rehearsal: true
		})

		expect(getPeripheralDeviceCommands(playoutDevice)).toHaveLength(1)
		expect(getPeripheralDeviceCommands(playoutDevice)[0]).toMatchObject({
			functionName: 'devicesMakeReady'
		})
	})
})
