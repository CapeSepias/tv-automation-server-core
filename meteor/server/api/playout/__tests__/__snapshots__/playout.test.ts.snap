// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playout API Basic rundown control 1`] = `
Array [
  Object {
    "_id": "mockStudio4_randomId10002_status",
    "classes": Array [
      "rundown_active",
    ],
    "content": Object {
      "deviceType": 0,
    },
    "enable": Object {
      "while": 1,
    },
    "id": "randomId10002_status",
    "layer": "rundown_status",
    "objectType": "rundown",
    "playlistId": "playlist_randomId10002",
    "rundownId": "randomId10002",
    "studioId": "mockStudio4",
  },
  Object {
    "_id": "mockStudio4_part_group_randomId10002_part0_0",
    "children": Array [],
    "content": Object {
      "deviceType": 0,
      "type": "group",
    },
    "enable": Object {
      "start": "now",
    },
    "id": "part_group_randomId10002_part0_0",
    "isGroup": true,
    "isPartGroup": true,
    "layer": "",
    "objectType": "rundown",
    "playlistId": "playlist_randomId10002",
    "priority": 5,
    "rundownId": "randomId10002",
    "studioId": "mockStudio4",
  },
  Object {
    "_id": "mockStudio4_part_group_firstobject_randomId10002_part0_0",
    "classes": Array [],
    "content": Object {
      "callBack": "partPlaybackStarted",
      "callBackData": Object {
        "partId": "randomId10002_part0_0",
        "rundownId": "randomId10002",
      },
      "callBackStopped": "partPlaybackStopped",
      "deviceType": 0,
      "type": "callback",
    },
    "enable": Object {
      "start": 0,
    },
    "id": "part_group_firstobject_randomId10002_part0_0",
    "inGroup": "part_group_randomId10002_part0_0",
    "layer": "group_first_object",
    "objectType": "rundown",
    "playlistId": "playlist_randomId10002",
    "rundownId": "randomId10002",
    "studioId": "mockStudio4",
  },
  Object {
    "_id": "mockStudio4_statObj",
    "content": Object {
      "deviceType": 0,
      "objCount": 3,
      "type": "nothing",
    },
    "enable": Object {
      "start": 0,
    },
    "id": "statObj",
    "layer": "__stat",
    "objectType": "stat",
    "studioId": "mockStudio4",
  },
]
`;

exports[`Playout API Basic rundown control 2`] = `
Object {
  "_id": "randomId10002",
  "_rank": 0,
  "dataSource": "mock",
  "externalId": "MOCK_RUNDOWN",
  "importVersions": Object {
    "blueprint": "",
    "core": "",
    "showStyleBase": "",
    "showStyleVariant": "",
    "studio": "",
  },
  "name": "Default Rundown",
  "peripheralDeviceId": "mockDevice5",
  "playlistId": "playlist_randomId10002",
  "showStyleBaseId": "randomId10000",
  "showStyleVariantId": "randomId10001",
  "studioId": "mockStudio4",
}
`;

exports[`Playout API Basic rundown control 3`] = `
Array [
  Object {
    "_id": "mockStudio4_statObj",
    "content": Object {
      "deviceType": 0,
      "objCount": 1,
      "type": "nothing",
    },
    "enable": Object {
      "start": 0,
    },
    "id": "statObj",
    "layer": "__stat",
    "objectType": "stat",
    "studioId": "mockStudio4",
  },
  Object {
    "_id": "mockStudio4_baseline_version",
    "content": Object {
      "deviceType": 0,
    },
    "enable": Object {
      "start": 0,
    },
    "id": "baseline_version",
    "layer": "baseline_version",
    "metadata": Object {
      "versions": Object {
        "blueprintId": "mockBlueprint0",
        "blueprintVersion": "0.0.0",
        "core": "0.0.0-test",
        "studio": "asdf",
      },
    },
    "objectType": "rundown",
    "playlistId": "",
    "rundownId": "",
    "studioId": "mockStudio4",
  },
]
`;

exports[`Playout API Basic rundown control 4`] = `
Object {
  "_id": "playlist_randomId10002",
  "active": false,
  "currentPartId": null,
  "externalId": "MOCK_RUNDOWNPLAYLIST",
  "holdState": 0,
  "name": "Default RundownPlaylist",
  "nextPartId": null,
  "nextPartManual": false,
  "nextTimeOffset": null,
  "peripheralDeviceId": "mockDevice5",
  "previousPartId": null,
  "rehearsal": false,
  "studioId": "mockStudio4",
}
`;

exports[`Playout API Basic rundown control 5`] = `
Object {
  "_id": "randomId10002",
  "_rank": 0,
  "dataSource": "mock",
  "externalId": "MOCK_RUNDOWN",
  "importVersions": Object {
    "blueprint": "",
    "core": "",
    "showStyleBase": "",
    "showStyleVariant": "",
    "studio": "",
  },
  "name": "Default Rundown",
  "peripheralDeviceId": "mockDevice5",
  "playlistId": "playlist_randomId10002",
  "showStyleBaseId": "randomId10000",
  "showStyleVariantId": "randomId10001",
  "studioId": "mockStudio4",
}
`;
